### PR TITLE
refactor: consolidate MempoolAcceptResult processing

### PR DESCRIPTION
**Pull request by glozow**
Every time we try to `ProcessTransaction` (i.e. submit a tx to mempool), we use the result to update a few net processing data structures. For example, after a failure, the {wtxid, txid, both, neither} (depending on reason) should be cached in `m_recent_rejects` so we don't try to download/validate it again.

There are 2 current places and at least 1 future place where we need to process `MempoolAcceptResult`:
- In the `ProcessMessage` logic after receiving and validating a tx
- In `ProcessOrphanTx` where we retry orphans
- With #28970, after processing a package of transactions, we should do these updates for each tx in the package.

Consolidate this code so it isn't repeated in 2 places and so we can reuse it in a future PR.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -582,6 +582,19 @@ class PeerManagerImpl final : public PeerManager
      */
     bool MaybeDiscourageAndDisconnect(CNode& pnode, Peer& peer);
 
+    /** Handle a transaction whose result was not MempoolAcceptResult::ResultType::VALID.
+     * @param[in]   was_rejected    Whether this tx has already been rejected before and is being
+     *                              reevaluated as an orphan. Used to avoid adding duplicate entries to
+     *                              vExtraTxnForCompact.
+     * Updates m_txrequest, m_recent_rejects, m_orphanage, and vExtraTxnForCompact. */
+    void ProcessInvalidTx(const CTransactionRef& tx, const TxValidationState& result, bool was_rejected, NodeId nodeid)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, g_msgproc_mutex, cs_main);
```
Comment (dergoegge) [inline src/net_processing.cpp:591]:
```suggestion
    void ProcessInvalidTx(NodeId nodeid, const CTransactionRef& tx, const TxValidationState& result, bool was_rejected)
        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, g_msgproc_mutex, cs_main);
```

nit: This is more inline with other method sigs in peerman.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -582,6 +582,19 @@ class PeerManagerImpl final : public PeerManager
...
+
+    /** Handle a transaction whose result was MempoolAcceptResult::ResultType::VALID.
+     * Updates m_txrequest, m_orphanage, and vExtraTxnForCompact. Also queues the tx for relay. */
+    void ProcessValidTx(const CTransactionRef& tx, NodeId nodeid, const std::list<CTransactionRef>& replaced_transactions)
```
Comment (dergoegge) [inline src/net_processing.cpp:595]:
```suggestion
    void ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, const std::list<CTransactionRef>& replaced_transactions)
```

File: src/net_processing.cpp [hunk 3]
```diff
@@ -582,6 +582,19 @@ class PeerManagerImpl final : public PeerManager
      */
     bool MaybeDiscourageAndDisconnect(CNode& pnode, Peer& peer);
 
+    /** Handle a transaction whose result was not MempoolAcceptResult::ResultType::VALID.
+     * @param[in]   was_rejected    Whether this tx has already been rejected before and is being
```
Comment (dergoegge) [inline src/net_processing.cpp:586]: Maybe `was_orphan` would be a better name?

**Review by dergoegge — COMMENTED** [review]
Concept ACK

just inline nits for now

File: src/net_processing.cpp (hunk 3 above)
Comment (glozow) [reply src/net_processing.cpp:586]: I was calling it this, but in the future it could also be `was_failed_for_too_low_fee`.

Comment (dergoegge) [reply src/net_processing.cpp:586]: Ok, how about `add_to_extra_compact_txs`?

**Comment by glozow** [comment]
Note to reviewers: this PR isn't supposed to change behavior at all. I have a [spreadsheet](https://docs.google.com/spreadsheets/d/1QFtQT1CB7-TagkDC2qwWubfYcoQHpAbJkkgdM9dy4TA/edit?usp=sharing) here enumerating what we do on master, and this PR should match that.
For tests, p2p_orphan_handling.py was written to hit some of these codepaths. Unit/fuzz tests for this are pretty difficult to achieve right now, but I would like to add some in the future after modularization - see #28031 (e.g.https://github.com/bitcoin/bitcoin/pull/28031/commits/ffa4d60bf11abac7aeeb9712e77245c67c15d3f6).

File: src/net_processing.cpp (hunk 1 above)
Comment (glozow) [reply src/net_processing.cpp]: done

File: src/net_processing.cpp (hunk 2 above)
Comment (glozow) [reply src/net_processing.cpp]: done

File: src/net_processing.cpp (hunk 3 above)
Comment (glozow) [reply src/net_processing.cpp:586]: Ok :handshake: `maybe_add_extra_compact_tx`

**Comment by instagibbs** [comment]
thanks for doing this! concept ACK

**Review by instagibbs — APPROVED** [review]
ACK 07cd510ffe791a4dfc1824c67fb440be780a4e2b

I can't find any behavior differences in this attempt, and it seems to be straight forwardly useful in the follow-up 1p1c PR.

Looking forward to making better unit test coverage; it's really quite lacking sadly.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -4276,24 +4301,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
...
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            // As this version of the transaction was acceptable, we can forget about any
-            // requests for it.
-            m_txrequest.ForgetTxHash(tx.GetHash());
-            m_txrequest.ForgetTxHash(tx.GetWitnessHash());
-            RelayTransaction(tx.GetHash(), tx.GetWitnessHash());
-            m_orphanage.AddChildrenToWorkSet(tx);
-
+            ProcessValidTx(pfrom.GetId(), ptx, result.m_replaced_transactions.value());
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/net_processing.cpp:4335]: Adding in the next PR, or will squash if I retouch :+1:

File: src/net_processing.cpp [hunk 5]
```diff
@@ -3085,53 +3151,18 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
...
-                // then we know that the witness was irrelevant to the policy
-                // failure, since this check depends only on the txid
-                // (the scriptPubKey being spent is covered by the txid).
-                // Add the txid to the reject filter to prevent repeated
-                // processing of this transaction in the event that child
-                // transactions are later received (resulting in
-                // parent-fetching by txid via the orphan-handling logic).
-                if (state.GetResult() == TxValidationResult::TX_INPUTS_NOT_STANDARD && porphanTx->HasWitness()) {
-                    // We only add the txid if it differs from the wtxid, to
-                    // avoid wasting entries in the rolling bloom filter.
```
Comment (achow101) [inline src/net_processing.cpp:3105]: In 07cd510ffe791a4dfc1824c67fb440be780a4e2b "[refactor] consolidate invalid MempoolAcceptResult processing"

This comment is lost.

**Review by achow101 — COMMENTED** [review]
ACK 07cd510ffe791a4dfc1824c67fb440be780a4e2b

**Review by dergoegge — APPROVED** [review]
Code review ACK 07cd510ffe791a4dfc1824c67fb440be780a4e2b

File: src/net_processing.cpp [hunk 6]
```diff
@@ -3037,6 +3046,63 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     return;
 }
 
+void PeerManagerImpl::ProcessInvalidTx(NodeId nodeid, const CTransactionRef& ptx, const TxValidationState& state,
```
Comment (TheCharlatan) [inline src/net_processing.cpp:3049]: Nit: The var names differ from the declaration for `CTransactionsRef&` and `TxValidationState&`. Is there a reason for the `p` in `ptx`?

**Review by sedited — APPROVED** [review]
ACK 07cd510ffe791a4dfc1824c67fb440be780a4e2b

File: src/net_processing.cpp (hunk 6 above)
Comment (glozow) [reply src/net_processing.cpp:3049]: `p` as in "pointer". I usually use `CMutableTransaction mtx` and `CTransaction tx`. Though I agree there's a mix...

**Pull request merged**